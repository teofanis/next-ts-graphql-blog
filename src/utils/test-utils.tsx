import { render, RenderOptions } from '@testing-library/react'
import { ReactElement } from 'react'
// Add in any providers here if necessary:

const Providers: React.FC = ({ children }) => {
  return <>{children}</>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
// override render method
export { customRender as render }
