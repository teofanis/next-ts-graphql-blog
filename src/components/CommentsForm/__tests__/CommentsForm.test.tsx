import { CommentsForm } from '@components/index'
import * as Services from '@services/index'
import { fireEvent, render, waitFor } from '@utils/test-utils'

const submitCommentMock = jest.fn(() =>
  Promise.resolve({
    createComment: () => false,
  })
)
const ServiceSpy = jest.spyOn(Services, 'submitComment')
ServiceSpy.mockImplementation(submitCommentMock)

describe('Comment Form Test', () => {
  it('should render properly', () => {
    const { container } = render(<CommentsForm slug={`some-slug`} />)
    expect(container).toMatchSnapshot()
  })

  it('requires all fields', () => {
    const { getByText } = render(<CommentsForm slug={`some-slug`} />)
    const button = getByText(/Post Comment/i)
    fireEvent.click(button)
    expect(getByText(/All Fields are required/i)).toBeInTheDocument()
  })

  it('it submits correctly', async () => {
    const { getByTestId, getByText } = render(
      <CommentsForm slug={`some-slug`} />
    )
    const button = getByTestId('submit-button')
    const textArea = getByTestId('comment-input')
    const Name = getByTestId('name-input')
    const Email = getByTestId('email-input')

    await waitFor(() => {
      fireEvent.change(textArea, { target: { value: 'some comment' } })
      fireEvent.change(Name, { target: { value: 'some name' } })
      fireEvent.change(Email, { target: { value: 'some email' } })
      fireEvent.click(button)
      expect(textArea).toHaveValue('some comment')
      expect(Services.submitComment).toHaveBeenCalled()
      expect(getByText(/Comment Submitted For Review!/i)).toBeInTheDocument()
    })
  })
})
