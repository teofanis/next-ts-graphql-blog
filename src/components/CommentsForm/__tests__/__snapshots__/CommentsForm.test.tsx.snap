// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment Form Test should render properly 1`] = `
<div>
  <div
    class="bg-white dark:bg-gray-500 shadow-lg rounded-lg p-8 pb-12 mb-8"
  >
    <h3
      class="text-xl mb-8 font-semibold border-b pb-4"
    >
      Leave a Comment
    </h3>
    <div
      class="grid grid-cols-1 gap-4 mb-4"
    >
      <textarea
        class="p-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
        data-testid="comment-input"
        name="comment"
        placeholder="Leave a comment"
      />
    </div>
    <div
      class="grid grid-cols-2 gap-4 mb-4"
    >
      <input
        class="py-2 px-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
        data-testid="name-input"
        name="name"
        placeholder="Name"
        type="text"
        value=""
      />
      <input
        class="py-2 px-4 outline-none w-full rounded-lg focus:ring-2 focus:ring-gray-200 bg-gray-100 text-gray-700"
        data-testid="email-input"
        name="email"
        placeholder="Email"
        type="email"
        value=""
      />
    </div>
    <div
      class="grid grid-cols-1 gap-4 mb-4"
    >
      <div>
        <input
          data-testid="checkbox-input"
          id="storeData"
          name="storeData"
          type="checkbox"
          value="true"
        />
        <label
          class="dark:text-white text-gray-500 cursor-pointer text-sm ml-2"
          for="storeData"
        >
          Save my e-email and name for the next time I comment.
        </label>
      </div>
    </div>
    <div
      class="mt-8"
    >
      <button
        class="transition duration-500 ease dark:hover:bg-yellow-600 dark:bg-yellow-500 hover:bg-indigo-900 inline-block bg-pink-600 text-lg rounded-full text-white px-8 py-3 cursor:pointer"
        data-testid="submit-button"
        type="button"
      >
        Post Comment
      </button>
    </div>
  </div>
</div>
`;
